### PR TITLE
Roll back queries on failure.

### PR DESCRIPTION
**Pull request by SamirTalwar**
### What

We need to end transactions if they fail. This ensures we do in the case of queries.

Mutations were already handled.

### How

I have moved the `rollback_on_exception` function to its own module and ensured that it is used in `query_engine_execution::query`.

To test this, I have constructed new metadata with just a couple of native queries: one broken, one working. The broken one is run first, then the working one afterwards. Before this change, the test failed because the working query would fail with the error:

```
error returned from database: current transaction is aborted, commands ignored until end of transaction block
```

### Limitations

Note that there are a couple of limitations with this implementation:

1. This behavior is optional. Adding more entrypoints into query execution will require us to remember to correctly handle transactions there too.
2. On a panic, the transaction will not be rolled back.

In the future, we may want to consider:

1. Delegating this behavior to `sqlx`. (We tried this but it meant an extra query to compensate for the `begin` function not taking transaction options, which degrades performance.)
2. Wrapping all database access to ensure we handle transactions correctly.
3. Handling panics.

### CODE REVIEW

**Review by soupi — COMMENTED** [review]
Looks good overall, but I have a few comments.

File: crates/tests/tests-common/src/request.rs [hunk 1]
```diff
@@ -3,26 +3,38 @@
...
 pub use axum::http::StatusCode;
-use ndc_client::models;
-use serde_derive::Deserialize;
+use axum::Router;
+use axum_test_helper::RequestBuilder;
+pub use axum_test_helper::TestClient;
+pub use ndc_client::models;
+
+/// Create a test client from a router.
+pub fn create_client(router: Router) -> TestClient {
```
[the remark this answers is not in the document]
Comment (SamirTalwar) [reply crates/tests/tests-common/src/request.rs:12]: It was used in a lot of places (and `axum_test_helper::` is a mouthful), so I decided to import them. I'll revert `Router` but leave `TestClient` as this module now re-exports that too.

File: static/empty-database.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+CREATE DATABASE empty;
```
[the remark this answers is not in the document]
Comment (SamirTalwar) [reply static/empty-database.sql:1]: I was building a fresh deployment from this database for the test too, but you made me realize that I don't need to do that, I can just use it directly. So yes, it's used for that one test case.

**Review by soupi — APPROVED** [review]
Thanks!

**Pull request merged**